File: python/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.top_node = None
        self.stack_size = 0

    def push(self, value):
        new_node = Node(value)
        new_node.next = self.top_node
        self.top_node = new_node
        self.stack_size += 1

    def pop(self):
        if self.empty():
            raise IndexError("Стек пуст")

        value = self.top_node.value
        self.top_node = self.top_node.next
        self.stack_size -= 1

        return value

    def top(self):
        if self.empty():
            raise IndexError("Стек пуст")

        return self.top_node.value

    def empty(self):
        return self.top_node is None

    def size(self):
        return self.stack_size

    def clear(self):
        self.top_node = None
        self.stack_size = 0

    def _print_from_bottom(self, node):
        if node is None:
            return

        self._print_from_bottom(node.next)

        if node.next is not None:
            print(" ", end="")

        print(node.value, end="")

    def print_stack(self):
        self._print_from_bottom(self.top_node)
        print()

File: cpp/src/stack.cpp
#include "stack.h"

#include <iostream>
#include <stdexcept>

Stack::Stack() : top_(nullptr), size_(0) {
}

Stack::~Stack() {
    clear();
}

void Stack::push(int value) {
    Node* newNode = new Node{value, top_};
    top_ = newNode;
    ++size_;
}

int Stack::pop() {
    if (empty()) {
        throw std::underflow_error("Stack is empty");
    }

    Node* oldTop = top_;
    int value = oldTop->value;

    top_ = oldTop->next;
    delete oldTop;

    --size_;

    return value;
}

int Stack::top() const {
    if (empty()) {
        throw std::underflow_error("Stack is empty");
    }

    return top_->value;
}

bool Stack::empty() const {
    return top_ == nullptr;
}

std::size_t Stack::size() const {
    return size_;
}

void Stack::clear() {
    while (top_ != nullptr) {
        Node* oldTop = top_;
        top_ = top_->next;
        delete oldTop;
    }

    size_ = 0;
}

void Stack::printFromBottom(Node* node, bool& first) const {
    if (node == nullptr) {
        return;
    }

    printFromBottom(node->next, first);

    if (!first) {
        std::cout << " ";
    }

    std::cout << node->value;
    first = false;
}

void Stack::print() const {
    bool first = true;
    printFromBottom(top_, first);
    std::cout << '\n';
}
